compiler/src/dotty/tools/dotc/interactive/Completion.scala

```diff
@@ -386,6 +365,61 @@ object Completion {
         !name.isConstructorName && name.toTermName.info.kind == SimpleNameKind
       def isStable = true
     }
+
+    object CompletionScope {
+      def from(denotations: Seq[SingleDenotation])(using Context): CompletionScope = {
+        val mappings = denotations
+          .filter(den => include(den.symbol, den.name))
+          .toList
+          .groupBy(_.name)
+          .map((name, denots) => name.stripModuleClassSuffix -> denots)
+        CompletionScope(mappings)
+      }
+
+      def fromNamed(namedDenotations: Seq[(Name, SingleDenotation)])(using Context): CompletionScope = {
+        val mappings = namedDenotations
+          .filter((name, denot) => include(denot.symbol, name))
+          .toList
+          .groupBy(_._1)
+          .map((name, namedDenots) => name.stripModuleClassSuffix -> namedDenots.map(_._2))
+        CompletionScope(mappings)
+      }
+
+      val empty = CompletionScope()
+    }
+
+    /** A wrapper over a map from names to their denotations
+     *  taking care of filtering out not matching completions
+     *  and enabling easy resolving of conflicts between different denotations with the same name
+     */
+    case class CompletionScope private(nameToDenots: Map[Name, List[SingleDenotation]] = Map.empty) {
+
+      // Merge two scopes using mappings from `that` instead of from `this` in case of name clashes
+      def mergeShadowedBy(that: CompletionScope) = CompletionScope(this.nameToDenots ++ that.nameToDenots)
+
+      // Merge two scopes but discard mappings for names that appear in both scopes
+      def mergeDiscardingAmbiguities(that: CompletionScope) =
+        val mappings = (this.nameToDenots.toList ++ that.nameToDenots.toList)
+          .groupBy(_._1)
+          .collect {
+            case (name, (_, denots) :: Nil)  => name -> denots
+          }
+        CompletionScope(mappings)
+
+      /**
+       * Return the list of symbols that should be included in completion results.
+       *
+       * If several symbols share the same name, the type symbols appear before term symbols inside
+       * the same `Completion`.
+       */
+      def getCompletions(using Context): List[Completion] = {
+        nameToDenots.toList.groupBy(_._1.toTermName.show).map { (name, namedDenots) =>
```

Comment:
```suggestion
        nameToDenots.toList.groupBy(_._1.toTermName).map { (name, namedDenots) =>
```

Name comparison is much faster than String.

Just curious, can we simply normalize the keys to TermName beforehand? This is done previously in `RenameAwareScope`.